### PR TITLE
make graceful shutdown thread-safe

### PR DESCRIPTION
**Pull request by ccl0326**

### CODE REVIEW

**Comment by benoitc** [comment]
+1 can anyone else review it? So far looks good for a merge

**Comment by benoitc** [comment]
how can I reproduce the issue?

**Comment by benoitc** [comment]
Sorry I didn't see the related issue #1032 . looking at it.

File: gunicorn/workers/gtornado.py [hunk 1]
```diff
@@ -50,10 +49,27 @@ def watchdog(self):
             self.log.info("Parent changed, shutting down: %s", self)
             self.stop()
```
Comment (fatelei) [inline gunicorn/workers/gtornado.py]: `stop` method is not defined. You have remove the method `stop`.

**Comment by ccl0326** [comment]
@fatelei Thanks.I'm so sorry for missed it out.Parent changed and Auto-restart has been tested.

File: gunicorn/workers/gtornado.py [hunk 2]
```diff
@@ -48,12 +46,29 @@ def watchdog(self):
 
         if self.ppid != os.getppid():
             self.log.info("Parent changed, shutting down: %s", self)
-            self.stop()
+            self.alive = False
+
+    def heartbeat(self):
+        if not self.alive:
+            if self.server_alive:
+                if hasattr(self, 'server'):
```
Comment (berkerpeksag) [inline gunicorn/workers/gtornado.py]: Can't we write this like the following snippet? (untested)

```py
if hasattr(self, 'server'):
    try:
        self.server.stop()
    except Exception:
        pass

self.server_alive = False
```

Comment (benoitc) [inline gunicorn/workers/gtornado.py]: indeed, if the server doesn't exist we need to stop the worker anyway.

Comment (ccl0326) [inline gunicorn/workers/gtornado.py]: The exception is ignored,but in fact we don't know how to deal with it.LGTM.

**Comment by berkerpeksag** [comment]
LGTM, thanks!

Two minor comments:

* Could you please squash the commits?
* Please delete the empty lines in gunicorn/workers/gtornado.py:

  ```diff
-# -*- coding: utf-8 -
+
+
   ```

**Comment by ccl0326** [comment]
done.

**Pull request merged**